### ISSUE TITLE
Reproducing on IMDB

### ISSUE THREAD

**Issue by cbbjames** [jhlau/topically-driven-language-model#7]
HI, thanks for the great work.

I'm trying to reproduce it on IMDB dataset.
However, I stuck in a high perplexity of testing:

test topic model perplexity = 16159.916
test language model perplexity = 77.684

I fixed some interface to make the code compatible with recent tensorflow:
https://github.com/cbbjames/topically-driven-language-model

The setting I used in config is trying to follow the paper:

vocab_minfreq=10 #minimum vocab frequency to filter
vocab_maxfreq=0.1 #proportion of most-frequent vocab to filter
stopwords="NLTKstopwords.txt"
tm_sent_len=3 #m_1; topic model sequence length
lm_sent_len=30 #m_2; language model sequence length
doc_len=150 #m_3; document max length

#training options
seed=1
batch_size=64 #n_batch
rnn_layer_size=1 #n_layer
rnn_hidden_size=600 #n_hidden
epoch_size=10 #n_epoch
topic_number=100 #k
word_embedding_size=300 #e; setting ignored if word_embeding_model is provided
word_embedding_model="word2vec/skipgram.bin" #pre-trained word embedding (gensim format); None if no pre-trained model
word_embedding_update=True #update word embedding for topic model
filter_sizes=[2] #h
filter_number=20 #a; topic input vector dimension
conv_activation="identity" #relu or identity (identity function is used in paper)
topic_embedding_size=50 #b; topic output vector dimension
learning_rate=0.001 #l
tm_keep_prob=0.4 #p_1
lm_keep_prob=0.6 #p_2
max_grad_norm=5 #gradient clipping
alpha=0.0 #additional loss to penalise similar topics; not used in paper (0.0)
num_samples=0 #sampled softmax to speed up training; not used in paper (0)
tag_embedding_size=0 #tag embedding dimension; 0 to disable tags

Could you shed some lights for me?
Thanks in advance!

**Comment by jhlau** [comment]
The hyper-parameters look fine; I am guessing when you upgraded the tensorflow version something (internally) may be implemented differently and resulted in the poor models.

Any chance you can train the original code using tf0.12? That can at least confirm that the problem is in the ported code.

**Comment by cbbjames** [comment]
Thanks for your reply.

I run it with tf 0.12 but the results are still far from the reports:

test topic model perplexity = 12188.849
test language model perplexity = 74.962

I'm also curious about:
1. 
could you kindly provide the stopword list in Sec. 4 in paper? 
I do observe that it significantly affects the results of evaluation. 

2.
In tdlm_model.py line 219:
 z, r = array_ops.split(1, 2, linear([self.conv_hidden, lstm_hidden],  2 * config.rnn_hidden_size, True, 1.0))
I wonder if that is the identical operation of Eq.3 in paper.
It seems to incorporate them into a single layer and split the outputs later, does that affects on the gate operations?

**Comment by jhlau** [comment]
You'll need to remove stopwords for the topic model, or it'll just learn a bunch of topics with stopwords as its top words.

The stopword list we used is detailed footnote 7: "We use Mallet’s stopword list: https://github.
com/mimno/Mallet/tree/master/stoplists"

**Comment by cbbjames** [comment]
Thanks.

I don't know why but using Mallet’s stopword list I got:

test topic model perplexity = 45042.526
test language model perplexity = 84.257

I also check topic-sents.txt, both the sentences and topics look quite reasonable (not stopwords).
If someone have hints please kindly let me know :)

**Comment by jhlau** [comment]
Really not sure what's wrong, but just out of curiosity, can you post the topics here?

**Comment by cbbjames** [comment]
Sure!

[topics.txt](https://github.com/jhlau/topically-driven-language-model/files/2133066/topics.txt)
[topic-sents.txt](https://github.com/jhlau/topically-driven-language-model/files/2133067/topic-sents.txt)

**Comment by jhlau** [comment]
Hmm the topics don't look coherent at all, so that means the topic model isn't working (hence the large perplexity)...

**Comment by cbbjames** [comment]
OK thanks.

We like your model very much and try to implement it for a useful component of our model. 
We will keep on trying and let you know if it works.

If any suggestions or findings please kindly let me know :)
